sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationApplicationSettingPropertySource.java

```diff
@@ -26,80 +30,108 @@
-
-                    // Null in the case of failFast is false, will just skip entry.
-                    if (entry != null) {
-                        properties.put(key, entry);
-                    }
-                } else if (StringUtils.hasText(setting.getContentType())
-                        && JsonConfigurationParser.isJsonContentType(setting.getContentType())) {
-                    Map<String, Object> jsonSettings = JsonConfigurationParser.parseJsonSetting(setting);
-                    for (Entry<String, Object> jsonSetting : jsonSettings.entrySet()) {
-                        key = jsonSetting.getKey().trim().substring(keyFilter.length());
-                        properties.put(key, jsonSetting.getValue());
-                    }
-                } else {
-                    properties.put(key, setting.getValue());
-                }
+            processConfigurationSettings(replicaClient.listSettings(settingSelector), settingSelector.getKeyFilter(),
+                trim, keyVaultClientFactory);
+        }
+
+    }
+
+    protected void processConfigurationSettings(List<ConfigurationSetting> settings, String keyFilter,
+        List<String> trimStrings, AppConfigurationKeyVaultClientFactory keyVaultClientFactory)
+        throws JsonProcessingException {
+        for (ConfigurationSetting setting : settings) {
+            if (trimStrings == null && StringUtils.hasText(keyFilter)) {
+                trimStrings = new ArrayList<>();
+                trimStrings.add(keyFilter.substring(0, keyFilter.length() - 1));
+            }
+            String key = trimKey(setting.getKey(), trimStrings);
+
+            if (setting instanceof SecretReferenceConfigurationSetting) {
+                handleKeyVaultReference(key, (SecretReferenceConfigurationSetting) setting);
+            } else if (setting instanceof FeatureFlagConfigurationSetting
+                && FEATURE_FLAG_CONTENT_TYPE.equals(setting.getContentType())) {
+                handleFeatureFlag(key, (FeatureFlagConfigurationSetting) setting, trimStrings);
+            } else if (StringUtils.hasText(setting.getContentType())
+                && JsonConfigurationParser.isJsonContentType(setting.getContentType())) {
+                handleJson(setting, trimStrings);
+            } else {
+                properties.put(key, setting.getValue());
             }
         }
     }
 
+    void handleKeyVaultReference(String key, SecretReferenceConfigurationSetting setting) {
+        String entry = getKeyVaultEntry(keyVaultClientFactory, setting);
+
+        // Null in the case of failFast is false, will just skip entry.
+        if (entry != null) {
+            properties.put(key, entry);
+        }
+    }
+
+    void handleFeatureFlag(String key, FeatureFlagConfigurationSetting setting, List<String> trimStrings)
+        throws JsonProcessingException {
+        handleJson(setting, trimStrings);
+    }
+
+    void handleJson(ConfigurationSetting setting, List<String> trimStrings)
```

Comment:
Better name for the `trimStrings`, seems like we are only trimming the prefixes.